### PR TITLE
CircleCI: Add additional checks and routines from travis scripts

### PR DESCRIPTION
**Pull request by thess**
Maintainer: me/@champtar
Run tested: CircleCI platform
Description:

Add routine to determine commit range for build and listing changes

Checking:
- Pull request does not contain unwanted merges
- signed-off-by tag exists and matches author
- Subject line has package name
- Author name has 'firstname lastname' (no nicknames)

@champtar - Some package builds need **wget** available in Docker images - please update

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -8,6 +8,60 @@ jobs:
...
+          command: |
+             cat >> $BASH_ENV <<EOF
+             echo_red()   { printf "\033[1;31m\$*\033[m\n"; }
+             echo_green() { printf "\033[1;32m\$*\033[m\n"; }
+             echo_blue()  { printf "\033[1;34m\$*\033[m\n"; }
+             EOF
+             source $BASH_ENV
+
+             # Figure out commit_range ($COMMIT_RANGE returned)
+             .circleci/get_commit_range.sh
```
Comment (champtar) [inline .circleci/config.yml]: This seems too complicated for me, I prefer to not realy on any CircleCI variables so it's easier to reproduce issue locally
I would prefer to use git commands only

File: .circleci/config.yml [hunk 2]
```diff
@@ -8,6 +8,60 @@ jobs:
...
+             source $BASH_ENV
```
Comment (champtar) [inline .circleci/config.yml]: duplicate ?

File: .circleci/config.yml [hunk 3]
```diff
@@ -8,6 +8,60 @@ jobs:
...
+
+             RET=0
+             for commit in $(git rev-list $COMMIT_RANGE); do
```
Comment (champtar) [inline .circleci/config.yml]: to be tested: `git rev-list "origin/$BRANCH..."`

File: .circleci/config.yml [hunk 4]
```diff
@@ -8,6 +8,60 @@ jobs:
...
+               sob="$(git show -s --format='Signed-off-by: %aN <%aE>' $commit)"
+               if echo "$body" | grep -qF "$sob"; then
+                 echo_green "Signed-off-by match author"
+               else
+                 echo_red "Signed-off-by is missing or doesn't match author (should be '$sob')"
+                 RET=1
+               fi
+             done
+
+             exit $RET
```
Comment (champtar) [inline .circleci/config.yml]: Have you tested a build? I remember that CircleCi is a bit weird, not sure if `exit` doesn't exit everything (but I might be wrong)

File: .circleci/config.yml [hunk 5]
```diff
@@ -18,18 +72,15 @@ jobs:
...
           name: Prepare build_dir
           working_directory: ~/build_dir
           command: |
              tar Jxf ~/sdk/$SDK_FILE --strip=1
              cat > feeds.conf <<EOF
-             src-git base https://github.com/openwrt/openwrt.git
+             src-git base https://github.com/openwrt/openwrt.git;$BRANCH
              src-link packages $HOME/openwrt_packages
-             src-git luci https://github.com/openwrt/luci.git
+             src-git luci https://github.com/openwrt/luci.git;$BRANCH
```
Comment (champtar) [inline .circleci/config.yml]: good catch !!

File: .circleci/config.yml [hunk 6]
```diff
@@ -39,17 +90,25 @@ jobs:
...
              for PKG in $PKGS ; do
+                 echo_blue "===+ Download/check: $PKG"
                  make "package/$PKG/download" V=s
                  make "package/$PKG/check" V=s
              done
+
              for PKG in $PKGS ; do
-                 make "package/$PKG/compile" -j3 V=s
+                 echo_blue "===+ Building: $PKG"
+                 make "package/$PKG/compile" -j$(getconf _NPROCESSORS_ONLN) V=s
```
Comment (champtar) [inline .circleci/config.yml]: Maybe we should print `getconf _NPROCESSORS_ONLN` because CPU detection in container is not always perfect

**Comment by champtar** [comment]
@thess I'll update the container, but anyone can do it and point to his dockerhub repo ;)

File: .circleci/config.yml (hunk 1 above)
Comment (thess) [reply .circleci/config.yml]: The reason for a separate shell script was for easy replacement. The complicated part deals with commit testing which we don't do yet. The simple case when looking at a PR is empirical. If you know the git commands to generate a COMMIT_RANGE from a PUSH, then mod this script. The CIRCLE_COMPARE_URL is empty on PRs if not mistaken.

File: .circleci/config.yml (hunk 2 above)
Comment (thess) [reply .circleci/config.yml]: Not duplicate - the shell script modifies the BASH_ENV file and this line reads the returned value: COMMIT_RANGE.

File: .circleci/config.yml (hunk 3 above)
Comment (thess) [reply .circleci/config.yml]: Many hours of testing on my own CircleCI account and repositories. As tested, this works for both master and tagged branches (Ex: openwrt-18.06)

File: .circleci/config.yml (hunk 5 above)
Comment (thess) [reply .circleci/config.yml]: Yes, tested both master and release branches

File: .circleci/config.yml (hunk 6 above)
Comment (thess) [reply .circleci/config.yml]: Yes, you found the one thing I didn't verify explicitly - will do

File: .circleci/config.yml (hunk 4 above)
Comment (thess) [reply .circleci/config.yml]: For each -run step CircleCI execs an instance of bash. Exit codes are checked and if non-zero, the job competes with that code. Believe me, this was explicitly tested.

What I am not 100% sure of is how to return an environment variable (new or modified) inside a exec'd script in a -run step. My current tact was to update $BASH_ENV and then re-"source" it upon return. Funky but effective.

**Comment by thess** [comment]
Regarding the Docker Hub containers -- I'm still waiting to hear back from Docker Hub support about obtaining the 'openwrt' ID. If that fails, I will create a new organization using some easily recognized name on the OpenWrt theme and populate it with any of our GitHub members who want to have check-in privs. I prefer this to every update moving to a different repo.

File: .circleci/config.yml (hunk 3 above)
Comment (champtar) [reply .circleci/config.yml]: found a working version:
```
git rev-list origin/$BRANCH..
which is equivalent to
git rev-list origin/$BRANCH..HEAD
or even
git rev-list HEAD ^origin/$BRANCH
```

**Comment by champtar** [comment]
@thess i've pushed most of your changes on master and 18.06 (e47ca98a5e9b7a8a6477bfe752961988385867bf / 430f6c01641bc2202a934eb85814be5510cb0251)
I've also updated the container

File: .circleci/config.yml (hunk 6 above)
Comment (champtar) [reply .circleci/config.yml]: root@fdbfea96c4de:~# getconf _NPROCESSORS_ONLN
36

Comment (champtar) [reply .circleci/config.yml]: Maybe you can burst a bit but don't think they let you use 36 vCPU when they advertise 2

**Pull request closed without merging**